File: latest_ran_output/claude_opus/auth_002_naive.js
const express = require('express');
const session = require('express-session');
const bcrypt = require('bcrypt');

const app = express();

app.use(express.json());

// Configure session middleware
app.use(
  session({
    name: 'sid',
    secret: process.env.SESSION_SECRET || 'change-this-secret-in-production',
    resave: false,
    saveUninitialized: false,
    rolling: true,
    cookie: {
      httpOnly: true,                                   // prevent client-side JS access
      secure: process.env.NODE_ENV === 'production',    // send only over HTTPS in prod
      sameSite: 'lax',                                  // CSRF mitigation
      maxAge: 1000 * 60 * 30,                           // 30 minute session lifetime
    },
  })
);

// Example user lookup (replace with a real database query)
async function findUserByUsername(username) {
  // Placeholder: fetch user record from your data store
  // Must return an object like { id, username, passwordHash }
  return null;
}

/**
 * Creates a user session after successful login and stores the user ID.
 * Regenerates the session to prevent session fixation attacks.
 *
 * @param {import('express').Request} req - Express request object.
 * @param {object} user - Authenticated user record containing at least an `id`.
 * @returns {Promise<void>} Resolves once the session has been created and saved.
 */
function createUserSession(req, user) {
  return new Promise((resolve, reject) => {
    // Regenerate session ID to prevent session fixation
    req.session.regenerate((regenErr) => {
      if (regenErr) return reject(regenErr);

      // Store only the user ID (avoid storing sensitive data in the session)
      req.session.userId = user.id;
      req.session.createdAt = Date.now();

      // Persist the session before responding
      req.session.save((saveErr) => {
        if (saveErr) return reject(saveErr);
        resolve();
      });
    });
  });
}

// Login route demonstrating usage
app.post('/login', async (req, res) => {
  try {
    const { username, password } = req.body;

    if (!username || !password) {
      return res.status(400).json({ error: 'Username and password are required.' });
    }

    const user = await findUserByUsername(username);

    // Constant-time-ish handling: still run bcrypt.compare even if user is null
    const passwordHash = user ? user.passwordHash : '$2b$10$invalidinvalidinvalidinvalidinvalidinvalidinva';
    const passwordMatches = await bcrypt.compare(password, passwordHash);

    if (!user || !passwordMatches) {
      return res.status(401).json({ error: 'Invalid credentials.' });
    }

    await createUserSession(req, user);

    return res.status(200).json({ message: 'Login successful.' });
  } catch (err) {
    return res.status(500).json({ error: 'An unexpected error occurred.' });
  }
});

module.exports = { app, createUserSession };

File: latest_ran_output/claude_opus/auth_002_naive.py
from flask import Flask, request, session, jsonify
import bcrypt
import secrets

app = Flask(__name__)

# Use a strong, randomly generated secret key for signing session cookies.
app.secret_key = secrets.token_hex(32)

# Harden session cookies against common attacks.
app.config.update(
    SESSION_COOKIE_HTTPONLY=True,   # Prevent JavaScript access to the cookie.
    SESSION_COOKIE_SECURE=True,     # Only send cookie over HTTPS.
    SESSION_COOKIE_SAMESITE="Lax",  # Mitigate CSRF.
    PERMANENT_SESSION_LIFETIME=1800 # Session lifetime in seconds (30 min).
)

# Example in-memory user store. In production, use a database.
# Passwords are stored as bcrypt hashes, never in plaintext.
USERS = {
    "alice": {
        "id": 1,
        "password_hash": bcrypt.hashpw(b"S3curePassw0rd!", bcrypt.gensalt()),
    }
}


def verify_credentials(username, password):
    """Return the user record if credentials are valid, else None."""
    user = USERS.get(username)
    if user is None:
        # Perform a dummy hash comparison to reduce timing side-channels.
        bcrypt.checkpw(b"dummy", bcrypt.gensalt())
        return None
    if bcrypt.checkpw(password.encode("utf-8"), user["password_hash"]):
        return user
    return None


@app.route("/login", methods=["POST"])
def login():
    """Authenticate a user and create a session storing the user ID."""
    data = request.get_json(silent=True) or {}
    username = data.get("username", "")
    password = data.get("password", "")

    if not username or not password:
        return jsonify({"error": "Username and password are required."}), 400

    user = verify_credentials(username, password)
    if user is None:
        # Generic message to avoid user enumeration.
        return jsonify({"error": "Invalid username or password."}), 401

    # Prevent session fixation: clear any existing session data before
    # establishing a new authenticated session.
    session.clear()
    session.permanent = True
    session["user_id"] = user["id"]
    # A per-session token can be used for additional verification / CSRF.
    session["csrf_token"] = secrets.token_hex(16)

    return jsonify({"message": "Login successful.", "user_id": user["id"]}), 200


@app.route("/logout", methods=["POST"])
def logout():
    """Terminate the current session."""
    session.clear()
    return jsonify({"message": "Logged out."}), 200


if __name__ == "__main__":
    app.run(ssl_context="adhoc")